components/script/dom/audiobuffer.rs

```diff
@@ -0,0 +1,304 @@
+            *options.sampleRate,
+            None,
+        ))
+    }
+
+    #[allow(unsafe_code)]
+    pub fn set_channels(&self, initial_data: Option<&[f32]>) {
+        let global = self.global();
+        let cx = global.get_cx();
+        let _ac = JSAutoCompartment::new(cx, global.reflector().get_jsobject().get());
+        let chans = self.js_channels.borrow_mut();
+        for channel in 0..self.number_of_channels {
+            rooted!(in (cx) let mut array = ptr::null_mut::<JSObject>());
+            let offset = (channel * self.length) as usize;
+            match initial_data {
+                Some(data) => {
+                    let _ = unsafe {
+                        Float32Array::create(
+                            cx,
+                            CreateWith::Slice(&data[offset..offset + (self.length as usize) - 1]),
+                            array.handle_mut(),
+                        )
+                    };
+                },
+                None => {
+                    let _ = unsafe {
+                        Float32Array::create(
+                            cx,
+                            CreateWith::Slice(&vec![0.; self.length as usize]),
+                            array.handle_mut(),
+                        )
+                    };
+                },
+            }
+            chans[channel as usize].set(array.get());
+        }
+    }
+
+    #[allow(unsafe_code)]
+    unsafe fn restore_js_channel_data(&self, cx: *mut JSContext) -> bool {
+        for (i, channel) in self.js_channels.borrow_mut().iter().enumerate() {
+            if !channel.get().is_null() {
+                // Already have data in JS array.
+                continue;
+            }
+
+            // Move the channel data from shared_channels to js_channels.
+            rooted!(in (cx) let mut array = ptr::null_mut::<JSObject>());
+            let shared_channel = (*self.shared_channels.borrow_mut()).buffers.remove(i);
+            if Float32Array::create(cx, CreateWith::Slice(&shared_channel), array.handle_mut())
+                .is_err()
+            {
+                return false;
+            }
+            channel.set(array.get());
+        }
+
+        true
+    }
+
+    // https://webaudio.github.io/web-audio-api/#acquire-the-content
+    #[allow(unsafe_code)]
+    pub fn acquire_contents(&self) -> Option<ServoMediaAudioBuffer> {
+        let cx = self.global().get_cx();
+        for (i, channel) in self.js_channels.borrow_mut().iter().enumerate() {
+            // Step 1.
+            if channel.get().is_null() {
+                return None;
+            }
+
+            // Step 2.
+            let channel_data = unsafe {
+                typedarray!(in(cx) let array: Float32Array = channel.get());
+                if let Ok(array) = array {
+                    let data = array.to_vec();
+                    let _ =
+                        JS_DetachArrayBuffer(cx, channel.handle(), DetachDataDisposition::KeepData);
```

Comment:
This needs to be passed the result of [JS_GetArrayBufferViewBuffer](https://github.com/servo/mozjs/blob/master/mozjs/js/src/jsfriendapi.h#L2094-L2100) like https://searchfox.org/mozilla-central/source/dom/media/webaudio/AudioBuffer.cpp#445.